src/ocrd_models/ocrd_mets.py

```diff
@@ -608,48 +607,60 @@ def get_physical_pages(self, for_fileIds : Optional[str] = None, for_pageIds : O
-                                    any(p in self._page_cache[a] for p in page_attr_patterns) or \
-                                    any([isinstance(p, typing.Pattern) and p.fullmatch(attr_val) \
-                                        for p in page_attr_patterns \
-                                        for attr_val in self._page_cache[a]]
-                                    )))
-                        for attr_val in self._page_cache[attr].keys():
-                            if attr_val in page_attr_patterns or \
-                                    any([isinstance(p, typing.Pattern) and p.fullmatch(attr_val) for p in page_attr_patterns]):
-                                if return_divs:
-                                    ret.append(self._page_cache[attr][attr_val])
-                                else:
-                                    ret.append(attr_val)
-                    except StopIteration:
-                        log.debug(f"No pattern matches any keys of any of the _page_caches. patterns: {page_attr_patterns}")
-                else:
-                    # determine attr during iterating
-                    attr = None
+                    page_attr_patterns.append(pageId_token)
+            if not page_attr_patterns:
+                return []
+            if self._cache_flag:
+                # determine attr to look for before iterating
+                try:
+                    for pat in page_attr_patterns:
+                        # for attr in list(METS_PAGE_DIV_ATTRIBUTE):
+                        attr : METS_PAGE_DIV_ATTRIBUTE
+                        if isinstance(pat, str):
+                            attr = next(a for a in list(METS_PAGE_DIV_ATTRIBUTE) if pat in self._page_cache[a])
+                            cache_keys = [pat]
+                        elif isinstance(pat, list):
+                            attr = next(a for a in list(METS_PAGE_DIV_ATTRIBUTE) if pat[0] in self._page_cache[a])
+                            cache_keys = [v for v in pat if v in self._page_cache[attr]]
+                        elif isinstance(pat, typing.Pattern):
+                            attr = next(a for a in list(METS_PAGE_DIV_ATTRIBUTE) for v in self._page_cache[a] if pat.fullmatch(v))
+                            cache_keys = [v for v in self._page_cache[attr] if pat.fullmatch(v)]
+                        else:
+                            raise ValueError
+                        if return_divs:
+                            ret += [self._page_cache[attr][v] for v in cache_keys]
+                        else:
+                            ret += cache_keys
+                except StopIteration:
+                    log.debug(f"No pattern matches any keys of any of the _page_caches. patterns: {page_attr_patterns}")
+            else:
+                while page_attr_patterns:
+                    pat = page_attr_patterns.pop(0)
                     for page in self._tree.getroot().xpath(
```

Comment:
I am pretty sure this is unnecessarily inefficient: first, because we usually have more physical pages than page patterns, and second, because we can `del` the pattern matches while iterating pages.

Reply:
26478310